service/src/main/java/bio/terra/workspace/service/workspace/flight/removeuser/ClaimUserPrivateResourcesStep.java

```diff
@@ -74,4 +80,13 @@ public StepResult undoStep(FlightContext context) throws InterruptedException {
         workspaceUuid, userEmail, context.getFlightId());
     return StepResult.getStepResultSuccess();
   }
+
+  private static boolean shouldCleanupResource(
+      ControlledResource resource, boolean userCanWriteWorkspace, boolean userCanReadWorkspace) {
+    return switch (resource.getRequiredSamActionForPrivateResource()) {
+      case SamWorkspaceAction.READ -> !(userCanWriteWorkspace || userCanReadWorkspace);
+      case SamWorkspaceAction.WRITE -> !userCanWriteWorkspace;
+      default -> true;
+    };
+  }
```

Comment:
this function is the crux of the change to the `RemoveUserFromWorkspaceFlight`. This step used to claim all the private resources and clean them up in subsequent steps, each step checking if the user still has access. This PR introduces the decision encoded in this function, releases the claim if the resource is not subject to cleanup and removes still-has-access checks in later steps.